File: java/src/main/java/it/giannibombelli/wsc2026/common/domain/primitive/Timestamp.java
package it.giannibombelli.wsc2026.common.domain.primitive;

import it.giannibombelli.wsc2026.common.utils.Require;

import java.time.Instant;

public record Timestamp(Instant value) {
    public Timestamp {
        Require.requireArgument(value, "timestamp");
    }

    public static Timestamp now() {
        return new Timestamp(Instant.now());
    }

    public Timestamp plusSeconds(long seconds) {
        return new Timestamp(value.plusSeconds(seconds));
    }

    public Timestamp minusSeconds(long seconds) {
        return new Timestamp(value.minusSeconds(seconds));
    }

    public boolean isBefore(Timestamp other) {
        return value.isBefore(other.value);
    }

    public boolean isAfter(Timestamp other) {
        return value.isAfter(other.value);
    }

    public boolean isBeforeOrEqual(Timestamp other) {
        return !value.isAfter(other.value);
    }

    public boolean isAfterOrEqual(Timestamp other) {
        return !value.isBefore(other.value);
    }
}

File: typescript/src/common/domain/primitive/timestamp.ts
import { requireArgument } from '@/common/utils/requireArgument.js';
export class Timestamp {
  readonly value: Date;

  constructor(value: Date) {
    requireArgument(value, 'timestamp');
    this.value = value;
  }

  plusSeconds(seconds: number): Timestamp {
    return new Timestamp(new Date(this.value.getTime() + seconds * 1000));
  }

  minusSeconds(seconds: number): Timestamp {
    return new Timestamp(new Date(this.value.getTime() - seconds * 1000));
  }

  isBefore(other: Timestamp): boolean {
    return this.value.getTime() < other.value.getTime();
  }

  isAfter(other: Timestamp): boolean {
    return this.value.getTime() > other.value.getTime();
  }

  isBeforeOrEqual(other: Timestamp): boolean {
    return this.value.getTime() <= other.value.getTime();
  }

  isAfterOrEqual(other: Timestamp): boolean {
    return this.value.getTime() >= other.value.getTime();
  }

  static now(): Timestamp {
    return new Timestamp(new Date());
  }
}
